app/src/main/cpp/skyline/kernel/scheduler.cpp

```diff
@@ -0,0 +1,367 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#include <unistd.h>
+#include <common/signal.h>
+#include "types/KThread.h"
+#include "scheduler.h"
+
+namespace skyline::kernel {
+    Scheduler::CoreContext::CoreContext(u8 id, u8 preemptionPriority) : id(id), preemptionPriority(preemptionPriority) {}
+
+    Scheduler::Scheduler(const DeviceState &state) : state(state) {}
+
+    void Scheduler::SignalHandler(int signal, siginfo *info, ucontext *ctx, void **tls) {
+        if (*tls) {
+            const auto &state{*reinterpret_cast<nce::ThreadContext *>(*tls)->state};
+            state.scheduler->Rotate(false);
+            YieldPending = false;
+            state.scheduler->WaitSchedule();
+        } else {
+            YieldPending = true;
+        }
+    }
+
+    Scheduler::CoreContext &Scheduler::LoadBalance(const std::shared_ptr<type::KThread> &thread, bool alwaysInsert) {
+        std::lock_guard migrationLock(thread->coreMigrationMutex);
+        auto *currentCore{&cores.at(thread->coreId)};
+
+        if (!currentCore->queue.empty() && thread->affinityMask.count() != 1) {
+            // Select core where the current thread will be scheduled the earliest based off average timeslice durations for resident threads
+            // There's a preference for the current core as migration isn't free
+            size_t minTimeslice{};
+            CoreContext *optimalCore{};
+            for (auto &candidateCore : cores) {
+                if (thread->affinityMask.test(candidateCore.id)) {
+                    u64 timeslice{};
+
+                    if (!candidateCore.queue.empty()) {
+                        std::unique_lock coreLock(candidateCore.mutex);
```

Comment:
A lot of `lock_guard`s were inherited here from when the core mutex was `std::shared_mutex` rather than `std::mutex`, after profiling it was determined that the costs of locking a shared mutex in exclusive mode were far greater than paying the costs for the slightly higher serialization with a regular mutex.